### PR TITLE
Fix support for Npgsql6 provider

### LINKED ISSUE #2994: Npgsql 6 is not compatible
Npgsql 6 has [breaking changes](https://www.npgsql.org/doc/release-notes/6.0.html#timestamp-rationalization-and-improvements) regarding DateTime (timespan/timespantz) handling that renders it incompatible with current version of NHibernate. Currently the only way to make it work is to enable `EnableLegacyTimestampBehavior`:

```csharp
AppContext.SetSwitch("Npgsql.EnableLegacyTimestampBehavior", true);
```

### PR DESCRIPTION
**Pull request by hazzik**
Fixes #2994

### CODE REVIEW

File: src/NHibernate/Driver/NpgsqlDriver.cs [hunk 1]
```diff
@@ -65,19 +65,33 @@ public override IResultSetsCommand GetResultSetsCommand(Engine.ISessionImplement
...
+				throw new QueryException($"No type assigned to parameter '{name}'");
 
-			// Since the .NET currency type has 4 decimal places, we use a decimal type in PostgreSQL instead of its native 2 decimal currency type.
+			dbParam.ParameterName = FormatNameForParameter(name);
 			if (sqlType.DbType == DbType.Currency)
+			{
+				// Since the .NET currency type has 4 decimal places, we use a decimal type in PostgreSQL instead of its native 2 decimal currency type.
 				dbParam.DbType = DbType.Decimal;
+			}
+			else if (DriverVersionMajor < 6 || sqlType.DbType != DbType.DateTime)
```
Comment (hazzik) [inline src/NHibernate/Driver/NpgsqlDriver.cs:77]: Mr. Smarty Pants.

**Comment by hazzik** [comment]
Maybe we should consider porting this to 5.3

**Comment by fredericDelaporte** [comment]
Ok for targeting 5.3.x, unless we consider releasing 5.4 soon.

**Comment by hazzik** [comment]
Ok, I've rebased to 5.3.x

**Review by fredericDelaporte — APPROVED** [review]

File: src/NHibernate/Driver/NpgsqlDriver.cs (hunk 1 above)
Comment (olga-sonix) [reply src/NHibernate/Driver/NpgsqlDriver.cs:77]: Hi!
I am sorry, it seems theese changes leads to an error in some cases.

Versions:
- 	NHibernate 5.4.2
- 	Npgsql 7.0.2

The named query:
```
	<sql-query name="MyQuery" cache-mode="normal" read-only="true">
	  <query-param type="DateTime" name="DateOfSearch"/>
	  <return-scalar type="Int64" column="SomeID"/>
	  <![CDATA[select ...
	        from ...
	        where ((  :DateOfSearch is null ) or (( t.DT1 <= :DateOfSearch) and ( t.DT2 >= :DateOfSearch )))
	        and ... ;]]>
	</sql-query>
	</hibernate-mapping>
```
Call the query:
```
	DateTime? dateOfSearch = null; // set null value here -- !!!!
	query.SetParameter("DateOfSearch", dateOfSearch, NHibernateUtil.DateTime);
	var result = query.List<long>();
```
NHibernate log:
```
	2023-04-21 23:01:51,433 [1] DEBUG NHibernate.Type.DateTimeType - binding null to parameter: 0
	2023-04-21 23:01:51,433 [1] DEBUG NHibernate.Type.DateTimeType - binding null to parameter: 1
	2023-04-21 23:01:51,434 [1] DEBUG NHibernate.Type.DateTimeType - binding null to parameter: 2
	2023-04-21 23:01:51,445 [1] DEBUG NHibernate.SQL - select ...
	        from ...
	        where ((  :p0 is null ) or (( mmh.DT_Mount <= :p0) and ( mmh.DT_Dismount >= :p0 ))); :p0 = NULL [Type: Object (0:0:0)]  -- !!!!
	...
	2023-04-21 23:01:51,623 [1] ERROR NHibernate.AdoNet.AbstractBatcher - Could not execute query: select ...
```
dbParam.DBType is empty, and Npgsql sends Prepare command to PostgreSQL with no parameter type specified, but PostgreSQL also can not guess the type from this query and value.
It works correctly, if I set non-null parameter value.

Comment (fredericDelaporte) [reply src/NHibernate/Driver/NpgsqlDriver.cs:77]: That point has been fixed by#3299.

**Pull request merged**